src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/SMDH.java

```diff
@@ -0,0 +1,167 @@
+        smdh = ByteBuffer.allocate(source.length);
+        smdh.position(0);
+        smdh.put(source);
+        smdh.position(0);
+
+        region = parseRegion();
+        icon = parseIcon();
+        parseMeta();
+    }
+
+    private GameRegion parseRegion(){
+        GameRegion region;
+        smdh.position(META_REGION_OFFSET);
+
+        int regionMasks = smdh.get() & 0xFF;
+
+        final boolean japan = (regionMasks & 0x1) != 0;
+        final boolean northAmerica = (regionMasks & 0x2) != 0;
+        final boolean europe = (regionMasks & 0x4) != 0;
+        final boolean australia = (regionMasks & 0x8) != 0;
+        final boolean china = (regionMasks & 0x10) != 0;
+        final boolean korea = (regionMasks & 0x20) != 0;
+
+        final boolean taiwan = (regionMasks & 0x40) != 0;
+        if (northAmerica) {
+            region = GameRegion.NorthAmerican;
+        } else if (europe) {
+            region = GameRegion.Europe;
+        } else if (australia) {
+            region = GameRegion.Australia;
+        } else if (japan) {
+            region = GameRegion.Japan;
+            metaIndex = 0;
+        } else if (korea) {
+            metaIndex = 7;
+            region = GameRegion.Korean;
+        } else if (china) {
+            metaIndex = 6;
+            region = GameRegion.China;
+        } else if (taiwan) {
+            metaIndex = 6;
+            region = GameRegion.Taiwan;
+        } else {
+            region = GameRegion.None;
+        }
+
+        return region;
+    }
+
+    private void parseMeta(){
+
+        for (int i = 0; i < 12; i++){
+            smdh.position(META_OFFSET + (512*i) + 0x80);
+            byte[] data = new byte[0x100];
+            smdh.get(data);
+            title[i] = convertString(data).replaceAll("\n", " ");
+        }
+
+        for (int i = 0; i < 12; i++){
+            smdh.position(META_OFFSET + (512 * i) + 0x180);
+            byte[] data = new byte[0x80];
+            smdh.get(data);
+            publisher[i] = convertString(data);
+        }
+    }
+
+    private int[] parseIcon() {
+        int[] icon = new int[ICON_SIZE*ICON_SIZE];
+        smdh.position(0);
+
+        for (int x = 0; x < ICON_SIZE; x++) {
+            for (int y = 0; y < ICON_SIZE; y++) {
+                int curseY = y & ~7;
+                int curseX = x & ~7;
+
+                int i = mortonInterleave(x, y);
+                int offset = (i + (curseX * 8)) * 2;
+
+                offset = offset + curseY * 48 * 2;
+
+                smdh.position(offset + IMAGE_OFFSET);
+
+                int bit1 = smdh.get() & 0xFF;
+                int bit2 = smdh.get() & 0xFF;
+
+                int pixel = bit1 + (bit2 << 8);
+
+                int r = (((pixel & 0xF800) >> 11) << 3);
+                int g = (((pixel & 0x7E0) >> 5) << 2);
+                int b = (((pixel & 0x1F)) << 3);
+
+                //Convert to ARGB8888
+                icon[x + 48 * y] = 255 << 24 | (r & 255) << 16 | (g & 255) << 8 | (b & 255);
```

Comment:
48 should be ICON_SIZE, 255 should be 0xff to show it's a bitmask, this is terrible to read without parenthesis and incredibly easy to get bugs like this. Likewise refer to textures.cpp